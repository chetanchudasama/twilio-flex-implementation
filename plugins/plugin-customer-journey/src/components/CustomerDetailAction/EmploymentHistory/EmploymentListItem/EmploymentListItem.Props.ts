import { CustomerEmploymentModel } from "@common/components";

export interface EmploymentListItemProps {
  employment: CustomerEmploymentModel;
  setActiveEmployment: () => void;
}
