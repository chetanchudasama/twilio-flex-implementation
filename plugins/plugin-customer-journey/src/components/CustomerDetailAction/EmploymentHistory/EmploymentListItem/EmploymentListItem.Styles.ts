import { paddingUnit, darkGrey, lightPurple } from "@common/components";
import { Grid } from "@material-ui/core";
import { styled } from "@twilio/flex-ui";

export const EmploymentListItemStyles = styled(Grid)({
  "&:last-child": {
    "& .divider": {
      display: "none",
    },
    "& .card-header": {
      paddingBottom: 0,
    },
  },
  ".icon": {
    backgroundColor: lightPurple,
  },
  ".title": {
    fontWeight: 500,
  },
  ".title-line-1": {
    fontSize: 16,
  },
  ".title-line-2": {
    fontSize: 14,
  },
  ".card-header": {
    alignItems: "flex-start",
    paddingLeft: 0,
    paddingRight: 0,
  },
  ".subheader": {
    paddingTop: paddingUnit * 2,
  },
  ".subheader-line": {
    color: darkGrey,
  },
  ".action-btn": {
    fontWeight: "bold",
  },
  ".action-icon": {
    marginRight: paddingUnit,
  },
});
