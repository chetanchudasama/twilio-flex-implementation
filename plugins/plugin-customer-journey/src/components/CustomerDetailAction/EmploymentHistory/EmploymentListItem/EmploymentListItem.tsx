import React, { useMemo } from "react";

import {
  Avatar,
  Button,
  CardHeader,
  Divider,
  Typography,
} from "@material-ui/core";
import BusinessCenterIcon from "@material-ui/icons/BusinessCenter";
import EditIcon from "@material-ui/icons/Edit";

import { EmploymentListItemProps } from "./EmploymentListItem.Props";
import { EmploymentListItemStyles } from "./EmploymentListItem.Styles";

export const EmploymentListItem: React.FC<EmploymentListItemProps> = ({
  employment,
  setActiveEmployment,
}) => {
  const occupationStatus = useMemo(() => {
    return employment.occupationStatus.name
      ? `(${employment.occupationStatus.name})`
      : "";
  }, [employment.occupationStatus]);

  const addressLine1 = useMemo(() => {
    const { employerAddress } = employment;
    const addressLine: string[] = [];
    if (employerAddress.buildingNumber) {
      addressLine.push(employerAddress.buildingNumber.toString());
    }
    if (employerAddress.buildingName) {
      addressLine.push(employerAddress.buildingName);
    }
    return addressLine.join(", ");
  }, [employment]);

  const addressLine2 = useMemo(() => {
    const { employerAddress } = employment;
    const addressLine: string[] = [];
    if (employerAddress.subBuilding) {
      addressLine.push(employerAddress.subBuilding);
    }
    if (employerAddress.streetName) {
      addressLine.push(employerAddress.streetName);
    }
    return addressLine.join(", ");
  }, [employment]);

  const addressLine3 = useMemo(() => {
    const { employerAddress } = employment;
    const addressLine: string[] = [];
    if (employerAddress.town) {
      addressLine.push(employerAddress.town);
    }
    if (employerAddress.postcode) {
      addressLine.push(employerAddress.postcode);
    }
    return addressLine.join(", ");
  }, [employment]);

  return (
    <EmploymentListItemStyles item xs={12}>
      <CardHeader
        avatar={
          <Avatar color="secondary" classes={{ colorDefault: "icon" }}>
            <BusinessCenterIcon color="secondary" />
          </Avatar>
        }
        title={
          <>
            <Typography component="span" className="title title-line-1">
              {[employment.occupation, occupationStatus].join(" ")}
            </Typography>
            <Typography component="span" className="title title-line-2">
              {employment.employerName}
            </Typography>
          </>
        }
        subheader={
          <>
            <Typography component="span" className="subheader-line">
              {`${addressLine1} ${addressLine2}`}
            </Typography>
            <Typography component="span" className="subheader-line">
              {addressLine3}
            </Typography>
            <Typography component="span" className="subheader-line">
              {employment.employerPhoneNumber}
            </Typography>
            <Typography component="span" className="subheader-line">
              {`${employment.yearsAtEmployment} years, ${employment.monthsAtEmployment} months`}
            </Typography>
          </>
        }
        action={
          <Button
            variant="text"
            size="small"
            color="secondary"
            className="action-btn"
            onClick={setActiveEmployment}
          >
            <EditIcon
              color="inherit"
              fontSize="small"
              className="action-icon"
            />
            Edit
          </Button>
        }
        classes={{
          root: "card-header",
          subheader: "subheader",
        }}
      />
      <Divider className="divider" />
    </EmploymentListItemStyles>
  );
};
