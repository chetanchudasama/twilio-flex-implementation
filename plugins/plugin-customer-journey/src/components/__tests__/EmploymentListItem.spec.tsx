import React from "react";
import { create } from "react-test-renderer";
import { mount } from "enzyme";
import { EmploymentListItem } from "../CustomerDetailAction/EmploymentHistory/EmploymentListItem/EmploymentListItem";
import { EmploymentListItemProps } from "../CustomerDetailAction/EmploymentHistory/EmploymentListItem/EmploymentListItem.Props";

const props: EmploymentListItemProps = {
  employment: {
    occupationStatus: {
      id: 1,
      name: "Full-time",
    },
    yearsAtEmployment: 2,
    monthsAtEmployment: 25,
    occupation: "Car rental agent",
    employerName: "Zara R Woods",
    employerPhoneNumber: "+446349900077",
    employerAddress: {
      postcode: "IP6 3YZ",
      buildingName: "Granary",
      subBuilding: "Granary-23",
      buildingNumber: "3",
      streetName: "75 Golf Road",
      town: "SWILLAND",
    },
    isCurrentEmployment: true,
  },
  setActiveEmployment: jest.fn(),
};

describe("EmploymentListItem", () => {
  it("renders correctly", () => {
    const wrapper = create(<EmploymentListItem {...props} />).toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it("should render employment detail correctly in UI", () => {
    const wrapper = create(<EmploymentListItem {...props} />).root;

    expect(
      wrapper.findByProps({
        className: "title title-line-1",
      }).props.children
    ).toContain(
      [
        props.employment.occupation,
        `(${props.employment.occupationStatus.name})`,
      ].join(" ")
    );
    expect(
      wrapper.findByProps({
        className: "title title-line-2",
      }).props.children
    ).toContain(props.employment.employerName);
  });

  it("when edit button is clicked, setActiveEmployment method should be emitted", () => {
    const wrapper = mount(<EmploymentListItem {...props} />);

    const button = wrapper.find(".action-btn").first();
    button.simulate("click");
    wrapper.update();

    expect(props.setActiveEmployment).toHaveBeenCalledTimes(1);
  });
});
