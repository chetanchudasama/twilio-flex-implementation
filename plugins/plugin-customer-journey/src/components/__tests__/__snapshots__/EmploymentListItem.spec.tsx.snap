// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmploymentListItem renders correctly 1`] = `
<div
  className="MuiGrid-item-2 MuiGrid-grid-xs-12-41 css-1es052e"
>
  <div
    className="MuiCardHeader-root-98 card-header"
  >
    <div
      className="MuiCardHeader-avatar-99"
    >
      <div
        className="MuiAvatar-root-104 MuiAvatar-colorDefault-105 icon"
        color="secondary"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-107 MuiSvgIcon-colorSecondary-109"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0zm10 5h4v2h-4zm0 0h4v2h-4z"
            fill="none"
          />
          <path
            d="M10 16v-1H3.01L3 19c0 1.11.89 2 2 2h14c1.11 0 2-.89 2-2v-4h-7v1h-4zm10-9h-4.01V5l-2-2h-4l-2 2v2H4c-1.1 0-2 .9-2 2v3c0 1.11.89 2 2 2h6v-2h4v2h6c1.1 0 2-.9 2-2V9c0-1.1-.9-2-2-2zm-6 0h-4V5h4v2z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiCardHeader-content-101"
    >
      <span
        className="MuiTypography-root-116 MuiTypography-body2-124 MuiCardHeader-title-102"
      >
        <span
          className="MuiTypography-root-116 MuiTypography-body1-125 title title-line-1"
        >
          Car rental agent (Full-time)
        </span>
        <span
          className="MuiTypography-root-116 MuiTypography-body1-125 title title-line-2"
        >
          Zara R Woods
        </span>
      </span>
      <span
        className="MuiTypography-root-116 MuiTypography-body2-124 MuiTypography-colorTextSecondary-149 MuiCardHeader-subheader-103 subheader"
      >
        <span
          className="MuiTypography-root-116 MuiTypography-body1-125 subheader-line"
        >
          3, Granary Granary-23, 75 Golf Road
        </span>
        <span
          className="MuiTypography-root-116 MuiTypography-body1-125 subheader-line"
        >
          SWILLAND, IP6 3YZ
        </span>
        <span
          className="MuiTypography-root-116 MuiTypography-body1-125 subheader-line"
        >
          +446349900077
        </span>
        <span
          className="MuiTypography-root-116 MuiTypography-body1-125 subheader-line"
        >
          2 years, 25 months
        </span>
      </span>
    </div>
    <div
      className="MuiCardHeader-action-100"
    >
      <button
        className="MuiButtonBase-root-178 MuiButton-root-152 MuiButton-text-154 MuiButton-textSecondary-156 MuiButton-flat-157 MuiButton-flatSecondary-159 MuiButton-sizeSmall-175 action-btn"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-153"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-107 MuiSvgIcon-fontSizeSmall-114 action-icon"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
          Edit
        </span>
        <span
          className="MuiTouchRipple-root-186"
        />
      </button>
    </div>
  </div>
  <hr
    className="MuiDivider-root-181 divider"
  />
</div>
`;
